### PR TITLE
Implement str.rsplit() based on CPython

### PR DESCRIPTION
**Pull request by densmirn**

### CODE REVIEW

**Comment by densmirn** [comment]
Could you please add the PR to review queue?

**Comment by stuartarchibald** [comment]
Thanks for the PR. Marked for review.

File: numba/unicode.py [hunk 1]
```diff
@@ -757,6 +758,121 @@ def split_whitespace_impl(a, sep=None, maxsplit=-1):
...
+
+    _unicode_rsplit_check_type(sep, 'sep', (types.UnicodeType,
+                                            types.UnicodeCharSeq,
+                                            types.NoneType))
+    _unicode_rsplit_check_type(maxsplit, 'maxsplit', (types.Integer, int))
+
+    if sep is None or isinstance(sep, (types.NoneType, types.Omitted)):
+
+        def rsplit_whitespace_impl(data, sep=None, maxsplit=-1):
+            # https://github.com/python/cpython/blob/master/Objects/stringlib/split.h#L192-L235    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode.py]: This needs an explicit git SHA reference.

File: numba/unicode.py [hunk 2]
```diff
@@ -757,6 +758,121 @@ def split_whitespace_impl(a, sep=None, maxsplit=-1):
...
+                    i -= 1
+                if i >= 0:
+                    result.append(data[0:i + 1])
+
+            return result[::-1]
+
+        return rsplit_whitespace_impl
+
+    def rsplit_impl(data, sep=None, maxsplit=-1):
+        # https://github.com/python/cpython/blob/master/Objects/stringlib/split.h#L286-L333    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode.py]: This needs an explicit git SHA reference.

File: numba/unicode.py [hunk 3]
```diff
@@ -757,6 +758,121 @@ def split_whitespace_impl(a, sep=None, maxsplit=-1):
...
+        if data._kind < sep._kind or len(data) < len(sep):
+            return [data]
+
+        def _rsplit_char(data, ch, maxsplit):
+            # https://github.com/python/cpython/blob/master/Objects/stringlib/split.h#L242-L284    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode.py]: This needs an explicit git SHA reference.

File: numba/unicode.py [hunk 4]
```diff
@@ -757,6 +758,121 @@ def split_whitespace_impl(a, sep=None, maxsplit=-1):
...
+            result = []
+            i = j = len(data) - 1
+            while i >= 0 and maxsplit > 0:
+                while i >= 0:
+                    data_code_point = _get_code_point(data, i)
+                    ch_code_point = _get_code_point(ch, 0)
```
Comment (stuartarchibald) [inline numba/unicode.py]: loop invariant, hoist?

File: numba/unicode_support.py [hunk 5]
```diff
@@ -311,6 +311,12 @@ def _PyUnicode_IsCaseIgnorable(ch):
     return ctype.flags & _PyUnicode_TyperecordMasks.CASE_IGNORABLE_MASK != 0
 
 
+@register_jitable
+def _PyUnicode_IsSpace(ch):
```
Comment (stuartarchibald) [inline numba/unicode_support.py]: Reference with SHA?

Comment (densmirn) [reply numba/unicode_support.py]: There is no such functionality in https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodectype.c. So I'm not able to insert a reference here.

**Comment by densmirn** [comment]
Could you please queue the PR for review?

File: numba/unicode_support.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/unicode_support.py]: Try this:
https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Include/unicodeobject.h#L125-L126
also, you'll note it has an ascii fast-path you may wish to implement.

**Comment by densmirn** [comment]
Could you please re-review the changes?

**Comment by stuartarchibald** [comment]
Please could you resolve conflicts against master via merge. Thanks.

**Comment by stuartarchibald** [comment]
Please could you resolve conflicts against master via merge. Thanks.

**Comment by densmirn** [comment]
@stuartarchibald all the conflicts resolved. So could you please continue reviewing the PR?

**Pull request closed without merging**